posthog/api/survey.py

```diff
@@ -751,6 +773,249 @@ def responses_count(self, request: request.Request, **kwargs):
+                }
+        return stats
+
+    def _calculate_rates(self, stats: SurveyStats) -> SurveyRates:
+        """Calculate response and dismissal rates from stats.
+
+        Args:
+            stats: Dictionary containing event stats
+
+        Returns:
+            Dictionary containing calculated rates
+        """
+        rates: SurveyRates = {
+            "response_rate": 0.0,
+            "dismissal_rate": 0.0,
+        }
+
+        shown_count = stats.get(SurveyEventName.SHOWN, {}).get("total_count", 0)
+        if shown_count > 0:
+            sent_count = stats.get(SurveyEventName.SENT, {}).get("total_count", 0)
+            dismissed_count = stats.get(SurveyEventName.DISMISSED, {}).get("total_count", 0)
+            rates = {
+                "response_rate": round(sent_count / shown_count * 100, 2),
+                "dismissal_rate": round(dismissed_count / shown_count * 100, 2),
+            }
+        return rates
+
+    def _get_survey_stats(self, date_from: str | None, date_to: str | None, survey_id: str | None = None) -> dict:
+        """Get survey statistics from ClickHouse.
+
+        Args:
+            date_from: Optional ISO timestamp for start date with timezone info
+            date_to: Optional ISO timestamp for end date with timezone info
+            survey_id: Optional survey ID to filter for. If None, gets stats for all surveys.
+
+        Returns:
+            Dictionary containing survey statistics and rates
+        """
+        parsed_from, parsed_to = self._validate_and_parse_dates(date_from, date_to)
+
+        # Build cache key based on parameters
+        cache_key_parts = [f"survey_stats_{self.team_id}"]
+        if survey_id:
+            cache_key_parts.append(str(survey_id))
+        # Use parsed dates for cache key to ensure consistent caching regardless of input timezone
+        cache_key_parts.extend([str(parsed_from), str(parsed_to)])
+        cache_key = "_".join(cache_key_parts)
+
+        # Try to get cached results first
+        cached_data = cache.get(cache_key)
+        if cached_data:
+            return cached_data
+
+        # Build query parameters
+        params: dict[str, Any] = {"team_id": str(self.team_id)}
+        date_filter = ""
+
+        if parsed_from:
+            date_filter += " AND timestamp >= %(date_from)s"
+            params["date_from"] = parsed_from
+        if parsed_to:
+            date_filter += " AND timestamp <= %(date_to)s"
+            params["date_to"] = parsed_to
+
+        # Add survey filter if specific survey
+        survey_filter = ""
+        if survey_id:
+            survey_filter = "AND JSONExtractString(properties, '$survey_id') = %(survey_id)s"
+            params["survey_id"] = str(survey_id)
+        else:
+            # For global stats, only include non-archived surveys
+            active_survey_ids = list(
+                Survey.objects.filter(team_id=self.team_id, archived=False).values_list("id", flat=True)
+            )
```

Comment:
it should be performant @ioannisj, clickhouse optimizes those queries internally. would only be a problem on the range of hundreds of thousands. hopefully we'll have a problme like that one day